app/src/main/java/io/homeassistant/companion/android/widgets/ButtonWidget.kt

```diff
@@ -0,0 +1,221 @@
+        // Every time AppWidgetManager.updateAppWidget(...) is called, the button listener
+        // and label need to be re-assigned, or the next time the layout updates
+        // (e.g home screen rotation) the widget will fall back on its default layout
+        // without any click listener being applied
+
+        val intent = Intent(context, ButtonWidget::class.java).apply {
+            action = CALL_SERVICE
+            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+        }
+
+        val views = RemoteViews(context.packageName, R.layout.widget_button).apply {
+            setOnClickPendingIntent(
+                R.id.widgetImageButton,
+                PendingIntent.getBroadcast(
+                    context,
+                    appWidgetId,
+                    intent,
+                    PendingIntent.FLAG_UPDATE_CURRENT
+                )
+            )
+            setTextViewText(
+                R.id.widgetLabel,
+                widgetStorage.loadLabel(appWidgetId)
+            )
+        }
+
+        return views
+    }
+
+    private fun callConfiguredService(context: Context, appWidgetId: Int) {
+        Log.d(TAG, "Calling widget service")
+
+        mainScope.launch {
+            // Change color of background image for feedback
+            val views = getWidgetRemoteViews(context, appWidgetId)
+            val appWidgetManager = AppWidgetManager.getInstance(context)
+
+            // Set default feedback as negative
+            var feedback = R.drawable.ic_circle_red_24dp
+
+            // Load the service call data from Shared Preferences
+            val domain = widgetStorage.loadDomain(appWidgetId)
+            val service = widgetStorage.loadService(appWidgetId)
+            val serviceData = widgetStorage.loadServiceData(appWidgetId)
+
+            Log.d(
+                TAG, "Service Call Data loaded:" + System.lineSeparator() +
+                        "domain: " + domain + System.lineSeparator() +
+                        "service: " + service + System.lineSeparator() +
+                        "service_data: " + serviceData
+            )
+
+            if (domain == null || service == null || serviceData == null) {
+                Log.w(TAG, "Service Call Data incomplete.  Aborting service call")
+            } else {
+                // If everything loaded correctly, package the service data and attempt the call
+                val serviceDataMap = HashMap<String, String>()
+                serviceDataMap["entity_id"] = serviceData
+
+                try {
+                    integrationUseCase.callService(domain, service, serviceDataMap)
+
+                    // If service call does not throw an exception, send positive feedback
+                    feedback = R.drawable.ic_circle_yellow_24dp
+                } catch (e: Exception) {
+                    Log.e(TAG, "Could not send service call.", e)
+                }
```

Comment:
One problem I had in other projects, is that if the Android device is on Power Save mode, all network requests from background fail.
To fix it, I needed to create a `ForegroundService` which then executes the network request on foreground.
You should probably test that

Reply:
Sorry, it may be unclear here due to the common use of the term "service" between Android and Home Assistant.  The method `callService(...)` is not actually calling an Android service, but is rather calling an available service within Home Assistant.  Specifically, it's using this API: https://developers.home-assistant.io/docs/en/app_integration_sending_data.html#call-a-service

If there is an issue with background data being suppressed within the IntegrationUseCase, I don't think it would be appropriate to attempt to fix that here, as that would be a deeper issue with a component to which I am adding but not significantly altering the core functionality.